### ISSUE TITLE
Auctionmark running for 24H in Azure

### ISSUE THREAD

**Issue by jarohen** [xtdb/crux#3478]
(24H is arbitrary - read 'long enough to flush out any immediate issues')

* Like #3477, but longer.
* Card up and attempt to minimise repros for any unearthed issues; we'll discuss the relative priorities of any larger tasks.
* Get an idea for the performance, how it compares to both your development machine and the local-only cloud run, how it behaves over time, etc.
* Generally, prioritise stability issues over performance unless the performance is show-stopping

**Comment by tggreene** [comment]
- We're switching to use AKS instead of container apps as the volume options for container apps are quite limited
- We have an AKS cluster with the benchmark running single node using workload identity a la [Workload Identity in AKS with Terraform](https://surajblog.medium.com/workload-identity-in-aks-with-terraform-9d6866b2bfa2).

**Comment by danmason** [comment]
### Notes on recent fixes/updates

- We saw a few issues very early on runs around restarting of nodes (ie, going from the load-phase to the 3 bench nodes)
  - We seem to have rectified the issues we noticed on the runs on restarts around IOOBEs via our work to ensure azure multiparts `abort` does not leave empty & unusable files #3629 
  - Fixed issues with closing behaviour on Kafka, #3680
 - We kept encountering heap size OoMs, issue number #3650 
   - Based on observations by myself and @tggreene, we noticed a high memory usage of the azure `getBlob` - essentially fetching arrow files into ByteBuffer memory prior to writing to disk.
   - Merged a fix to essentially download to a file directly and skip this/reduce the memory usage. 
   - See #3692 
   - Following merging & fixing up of the above we've yet to encounter any further Heap OoMs

### Current state of runs

Most recently, we ran a multinode run for twelve hours. Can see a folder with the logs of the various nodes on my fork, [**here**](https://github.com/danmason/xtdb/tree/12-hour-logs/cloud-benchmark/azure/azure-logs/12h).

We did indeed see one of the three nodes finishing "properly" - running for that amount of time and pinging a message within the slack channel. Some general observations based on previous issues:
- No more IOOBEs have been observed.
- No heap size OoMs have been observed.

Of the three nodes that ran - I'm only convinced that "node 2" within those logs above is the one that ran the full 12 hours/didn't fail out near the end/fail to complete. I do believe some of the auctionmarks OLTP threads fell over/failed within all of the nodes, based on the logs. Noting some general issues from across the three nodes:
- We see numerous warnings across the nodes from `reactor-http-epoll` pertaining to HttpClientConnect about connection errors when trying to communicate with Azure Blob Storage (xtdbazurebenchmark.blob.core.windows.net).
  - Sample warning and it's error stack included below.
  - Seemingly these get retried and pertain to the underlying Azure connection?
  - We've seen these many times across our various benchmarks/tests.
- Issues with running out of direct memory during compaction jobs.
  - In a few places we similar errors, such as "org.apache.arrow.memory.OutOfMemoryException: Failure allocating buffer.", called out from `merge-segments-into`.
  - Could be to do with simulatenous compaction jobs going above direct memory thresholds.
  - More complete error stack below. 
- Observed a **different** IndexOutOfBounds exception from one of the nodes.
  - Seems to be caused by `java.lang.IllegalStateException: RefCnt has gone negative` being thrown from an underlying arrow buffer.
  - Causes problem both when calling a query and within the local disk cache removal listener.
- One of the nodes seems to hang for quite a few hours / there are no running threads left. Can observe messages/activity at around 3am which stops up until the stagnant log flusher gets called and one of the thread throws an `InterruptedException`.
  - Could well be a result of other threads dying out due to various issues above.
  - Log included under error stacks. 
- Supplementing our notes on `TimeoutException` - one of the nodes did seem to stop doing work at a certain point and start outputting `java.util.concurrent.TimeoutException` periodically (and similar to the above, didn't complete properly).

As a side note - the logs are fairly noisy due to the "BlobAlreadyExists" message being `log/info` - we might want to drop this to `log/debug` given how common it seems to be for that to happen across threads. (Though it has been fairly useful as a quick check in the log that "stuff is happening on the node".

## Notable Error Stacks

### `WARN  r.n.http.client.HttpClientConnect`

A few different bits of networking issues captured in the above block - not sure if they're causing issues in practice. We might be able to update/tweak the retry/connection behaviours somewhat, though would have hoped we could have relied on Azure defaults for this. Perhaps also an issue of versioning.

```
20:51:33.687 [reactor-http-epoll-2] WARN  r.n.http.client.HttpClientConnect - [143b2e14-25, L:/10.244.1.124:46746 - R:xtdbazurebenchmark.blob.core.windows.net/20.60.197.33:443] The connection observed an error
java.util.concurrent.TimeoutException: Channel write operation timed out after 60000 milliseconds.
	at com.azure.core.http.netty.implementation.AzureSdkHandler.writeTimeoutRunnable(AzureSdkHandler.java:164)
	at com.azure.core.http.netty.implementation.AzureSdkHandler.lambda$startWriteTracking$0(AzureSdkHandler.java:100)
	at io.netty.util.concurrent.PromiseTask.runTask(PromiseTask.java:98)
	at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:159)
	at io.netty.util.concurrent.AbstractEventExecutor.runTask(AbstractEventExecutor.java:173)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:166)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:469)
	at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:408)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:994)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
20:51:36.705 [reactor-http-epoll-2] WARN  io.netty.handler.ssl.SslHandler - [id: 0x143b2e14, L:/10.244.1.124:46746 - R:xtdbazurebenchmark.blob.core.windows.net/20.60.197.33:443] Last write attempt timed out; force-closing the connection.
20:52:34.198 [reactor-http-epoll-2] WARN  r.n.http.client.HttpClientConnect - [a53f2bdc-31, L:/10.244.1.124:39108 ! R:xtdbazurebenchmark.blob.core.windows.net/20.60.197.33:443] The connection observed an error
reactor.netty.http.client.PrematureCloseException: Connection has been closed BEFORE response, while sending request body
20:52:35.846 [reactor-http-epoll-2] WARN  r.n.http.client.HttpClientConnect - [922f2ba4-33, L:/10.244.1.124:39094 - R:xtdbazurebenchmark.blob.core.windows.net/20.60.197.33:443] The connection observed an error, the request cannot be retried as the headers/body were sent
io.netty.channel.unix.Errors$NativeIoException: recvAddress(..) failed: Connection reset by peer
20:52:37.660 [reactor-http-epoll-2] WARN  r.n.http.client.HttpClientConnect - [38f972d0-32, L:/10.244.1.124:39124 ! R:xtdbazurebenchmark.blob.core.windows.net/20.60.197.33:443] The connection observed an error
reactor.netty.http.client.PrematureCloseException: Connection has been closed BEFORE response, while sending request body
20:52:37.671 [reactor-http-epoll-2] WARN  r.n.http.client.HttpClientConnect - [44fd7c32-29, L:/10.244.1.124:39232 - R:xtdbazurebenchmark.blob.core.windows.net/20.60.197.33:443] The connection observed an error, the request cannot be retried as the headers/body were sent
io.netty.channel.unix.Errors$NativeIoException: recvAddress(..) failed: Connection reset by peer
```

### `org.apache.arrow.memory.OutOfMemoryException: Failure allocating buffer.`

We've seen variations of the following a few times across our various benchmark runs - seems to kick in from `merge-segments-into`. I'm not sure if the compaction process completely falls over in this case, given the benchmark seems to continue regardless, but one to be aware of.

```
21:38:55.326 [xtdb.compactor-45] ERROR xtdb.compactor - Error running compaction job.
org.apache.arrow.memory.OutOfMemoryException: Failure allocating buffer.
	at io.netty.buffer.PooledByteBufAllocatorL.allocate(PooledByteBufAllocatorL.java:62)
	at org.apache.arrow.memory.netty.NettyAllocationManager.<init>(NettyAllocationManager.java:74)
	...
	at xtdb.compactor$__GT_reader__GT_copier$reader__GT_copier$reify__16301.copyRow(compactor.clj:50)
	at xtdb.compactor$merge_segments_into.invokeStatic(compactor.clj:92)
	at xtdb.compactor$merge_segments_into.invoke(compactor.clj:55)
	at xtdb.compactor$exec_compaction_job_BANG_$fn__16352$fn__16353$fn__16362$fn__16363.invoke(compactor.clj:136)
	at xtdb.compactor$exec_compaction_job_BANG_$fn__16352$fn__16353$fn__16362.invoke(compactor.clj:134)
	at xtdb.compactor$exec_compaction_job_BANG_$fn__16352$fn__16353.invoke(compactor.clj:134)
	at xtdb.compactor$exec_compaction_job_BANG_$fn__16352.invoke(compactor.clj:123)
      ...
Caused by: io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 268435456 byte(s) of direct memory (used: 1920991239, max: 2097152000)
	at io.netty.util.internal.PlatformDependent.incrementMemoryCounter(PlatformDependent.java:845)
	at io.netty.util.internal.PlatformDependent.allocateDirectNoCleaner(PlatformDependent.java:774)
	at io.netty.buffer.UnpooledUnsafeNoCleanerDirectByteBuf.allocateDirect(UnpooledUnsafeNoCleanerDirectByteBuf.java:30)
	...
```

### `java.lang.IllegalStateException: RefCnt has gone negative`

Believe we may have a card for this/a similar problem elsewhere, but seems to be showing up/causing threads to die within auctionmark.

```
22:57:09.402 [core2-benchmark-pool-7-thread-7] ERROR xtdb.bench - Error while executing xtdb.bench.auctionmark$proc_new_bid@4df4499a
java.lang.IndexOutOfBoundsException: index: -1, length: 1 (expected: range(0, 21))
	at org.apache.arrow.memory.ArrowBuf.checkIndexD(ArrowBuf.java:299)
	at org.apache.arrow.memory.ArrowBuf.chk(ArrowBuf.java:285)
	...
	at xtdb.node.impl.Node.open_sql_query(impl.clj:117)
	at xtdb.api$plan_q$reify__7747.reduce(api.clj:81)
	...
	Suppressed: java.lang.IllegalStateException: RefCnt has gone negative
		at org.apache.arrow.util.Preconditions.checkState(Preconditions.java:422)
		at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:128)
		at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:104)
		at org.apache.arrow.memory.ArrowBuf.close(ArrowBuf.java:1040)
		...
		at xtdb.util$close.invokeStatic(util.clj:56)
		at xtdb.util$close.invoke(util.clj:37)
		at xtdb.operator.scan$eval26459$fn$reify__26464$fn__26567$fn__26574$fn__26575$fn__26602.invoke(scan.clj:454)
		at xtdb.operator.scan$eval26459$fn$reify__26464$fn__26567$fn__26574$fn__26575.invoke(scan.clj:454)
		... 37 common frames omitted
Sep 12, 2024 10:57:09 PM com.github.benmanes.caffeine.cache.BoundedLocalCache lambda$notifyRemoval$1
WARNING: Exception thrown by removal listener

java.lang.IllegalStateException: RefCnt has gone negative
	at org.apache.arrow.util.Preconditions.checkState(Preconditions.java:422)
	at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:128)
	at org.apache.arrow.memory.BufferLedger.release(BufferLedger.java:104)
	at org.apache.arrow.memory.ArrowBuf.close(ArrowBuf.java:1040)
	at xtdb.util$close.invokeStatic(util.clj:54)
	at xtdb.util$close.invoke(util.clj:37)
	at xtdb.buffer_pool$$reify__8477.onRemoval(buffer_pool.clj:214)
	at com.github.benmanes.caffeine.cache.BoundedLocalCache.lambda$notifyRemoval$1(BoundedLocalCache.java:422)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1423)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1843)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1808)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:188)
```

### Node hanging up until end of run

Can see that it is logging out various calls up until a certain point where it goes quiet (other nodes continue outputting the "Blob already exists" messages for quite a while. Seems like at the very end as the OLTP process ends a few things get thrown/woken up.

```
01:43:20.074 [xtdb.compactor-352] INFO  xtdb.azure.object-store - Blob already exists for https://xtdbazurebenchmark.blob.core.windows.net/xtdbazurebenchmarkcontainer/xtdb-object-store%2Fv05%2Ftables%2Fpublic%24item_bid%2Fdata%2Flog-l01-fr541aac2-nr54b0ad3-rsccf5.arrow - aborting multipart upload
01:43:20.969 [xtdb.compactor-352] INFO  xtdb.azure.object-store - Blob already exists for xtdb-object-store/v05/tables/public$item_bid/meta/log-l01-fr541aac2-nr54b0ad3-rsccf5.arrow - terminating put operation
08:48:06.429 [xtdb.stagnant-log-flush-pool-6-thread-1] INFO  xtdb.stagnant-log-flusher - last chunk tx-id 754998, flushing any pending writes
08:54:46.998 [core2-benchmark-pool-8-thread-2] ERROR xtdb.bench - Error while executing xtdb.bench.auctionmark$index_item_status_groups@559a06a7
java.lang.InterruptedException: null
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:386)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at clojure.core$deref_future.invokeStatic(core.clj:2317)
	at clojure.core$deref.invokeStatic(core.clj:2337)
	at clojure.core$deref.invoke(core.clj:2323)
	at xtdb.indexer.Indexer.awaitTx(indexer.clj:563)
	at xtdb.node.impl.Node.prepareQuery(impl.clj:141)
	at xtdb.node.impl.Node.open_xtql_query(impl.clj:121)
	at xtdb.api$plan_q$reify__7747.reduce(api.clj:82)
	at clojure.core$reduce.invokeStatic(core.clj:6964)
```

**Comment by danmason** [comment]
Notes from planning:
- On the HTTP client errors/warnings, we would like to see if they ever get surfaced from our API calls to Azure.
  - Where we call down to the API, add extra handling of generic errors/exceptions and ensure we log, wrap and re-throw them ourselves.
  - When we're happy that these are being appropriately caught & handled within XTDB, want to update default logging (say on the benchmark + azure docker image logback files) to turn `reactor` and co up to `ERROR` logs only.
- Direct memory allocation errors:
  - Believe that based on the compaction job behaviour/handling of errors, would get retried/shouldn't cause the node to fall over. Want to verify this.
  - Want more generally prometheus metrics available to us - exposing those to whatever kubernetes/AKS expects/primarily uses 
    - Auctionmark has a whole load of gauges so we would like to use these to confirm that the benchmark is still alive.
    - We want to ensure that all of our running nodes are split up within wherever we expose the metrics - probably want to look at "dimensions" for this.
  - In arrow we use a lot of child allocators - each different component in XT has a child allocator associated with it, want to create a gauge for each of these child allocators to monitor node direct memory and where it is being allocated.
- `RefCnt has gone negative` error - I've pinged a comment with the stacktrace on our existing issue, https://github.com/xtdb/xtdb/issues/3551

Made some issues for the above:
- #3703
- #3701
- #3702 

Considering this "blocked" for now as we go towards a better logging & monitoring story and can better observe for issues.

**Comment by danmason** [comment]
Following us finishing up the various raised issues above, I've removed the `blocked` label.

With:
- Better error handling
- Application monitoring via Azure Application Insights
- Additional gauges for direct memory usage 
- Everything running within the same region

We should be in a good state to kick off further 24 hour runs and observe any further issues against `main`. Shall continue to monitor the results of these, and make note of all relevant behaviour.

**Comment by danmason** [comment]
# Current Overview

From a recent 24hour run (did not run for the full time), some observations:
- Had to fix some issues with port bindings (since we start a pgwire server by default now)
- Load phase ran through and completed just fine.
- Of the 3 nodes:
  - The first seemed to continue to run fine until the run was cancelled.
  - The second was OoM killed by Kubernetes after around an hour
  - The third continued to have a running process, though stopped doing any work after a few hours. Seems as if all active threads were blocked. 

Observing two of the three nodes as effectively closed, decided to cancel the run and head over to application insights. Within insights, also added a set of graphs to track the direct memory allocation gauges added within 63e3698.

## Application Insight Graphs

> **NOTE:** Within the graphs below, the dotted line represents areas where no metrics are being reported. 

### Memory usage per node
![image](https://github.com/user-attachments/assets/04cf1fff-a62d-42c8-ab62-b4f67aad8908)

### Auctionmark calls per node 

Can see here where exactly each node stops reporting (with bench-node-1 still running)
![image](https://github.com/user-attachments/assets/8664be2c-60df-460d-8aaf-9d98079bbc7d)

### Memory usage of Node 2 (which was OoM killed)

**Max/Average JVM Heap usage & total max buffer memory:**
![image](https://github.com/user-attachments/assets/0a39d4ef-b50d-4ac4-a12c-5b818d6af980)


**Direct child allocator max allocated memory:** 
![image](https://github.com/user-attachments/assets/7fe1f36c-b481-4230-9b7a-f8af5537eb32)

### Thread states of Node 3 (which seemed to hang after 9PM)

![image](https://github.com/user-attachments/assets/4f4998d5-4455-45fb-8397-d94289e338b0)


## Current questions

- Why did node 2 get OoM killed?
- Why did node 3 seem to just "stop" - no OoM killed, no errors of note, threads just stopped/got stuck waiting?
  - We could likely investigate this by running Yourkit profiler agents on all of the nodes and if the process DOES get stuck in this way, plugging in Yourkit and seeing what the nodes are doing.

**Comment by danmason** [comment]
Kicked off another 24 hour run to see if we could observe the process hanging problem with Yourkit running - have yet to see that, currently:
- Node 1 and Node 3 both got OOMKilled:
  - After about an hour of runtime.
  - Within about a minute of each other, by the look of it. :thinking_face:
- Node 2 is still running currently, doesn't seem to have hung.

## Graphs

## Node 1 Memory Usage (OoMKilled)
![image](https://github.com/user-attachments/assets/276890f3-2076-42d3-8adf-d6f367b70ef9)
![image](https://github.com/user-attachments/assets/1a1e31d9-649a-4f9c-8781-8e3e8d56ca4c)

## Node 2 Memory Usage
![image](https://github.com/user-attachments/assets/72b30a3f-52b2-42a1-a8df-8f73b86cb7b5)
![image](https://github.com/user-attachments/assets/9af36770-ced9-40dc-a283-052747d823ca)


## Node 3 Memory Usage (OoMKilled)
![image](https://github.com/user-attachments/assets/084d9688-f0e7-4bf4-837e-abf4b6d7d4e4)
![image](https://github.com/user-attachments/assets/77aa734e-f193-4b35-a710-2d3295cd84c8)

## Observations

- Node 3 does have some large spikes in buffer pool allocation - though still runs for a while before being OoM killed.
- Node 1 doesn't seem particularly high on memory usage either.
- In summary, not TOO sure about the OoM killing of pods.

# Edit: Followup

Node 2 was ALSO OoM killed - memory graphs at the point of that happening:
![image](https://github.com/user-attachments/assets/0666becc-42fd-4e36-8125-706027141055)
![image](https://github.com/user-attachments/assets/cd78ce37-a0f6-444f-89be-f9caf09fdb6c)

Have spun down the node pool for now.

**Comment by danmason** [comment]
New cards for both the OoMKilled Issue (#3769) and Process Hanging (#3770). Marking this as blocked for now.

**Comment by danmason** [comment]
Updating with recently observed behaviour against this:

## 26/11/2024

Kicked off a 24hour benchmark run on the 25th, observing the following:
- Node 1 hit into "no disk space left on device" - seems like we're overreaching our bounds in the pinned entry/disk cache.
  - Created a card, #3913. 
- Node 2 OoMKilled, #3769 
- Node 3 ran for 24 hours successfully.

Kicked off another 24hour run, doing the following to try and avoid the problems we observed:
- PersistentVolumeClaims to 100GiB each.
- Memory request on the nodes to 6GB each
- Running grafana alongside the node.

## 27/11/2024

As of 3:10PM (around 21 hours after the run started):
- All three nodes still running successfully.

At the end of the run:
- All three nodes ran successfully for 24hours!

This was with increased persistent volume claim size (the utilized size on disk was steadily increasing, as per #3913), and increased memory reservation. Kicked off another 24h run.

## 28/11/2024 

Run was still going with all three nodes after 19 hours - decided to cancel this run to test out my fix to #3913 - #3921.

New run kicked off with:
- Persistent volume claim size back down to 50GiB.
- Memory reservation still at 6GB.

Shall observe and see if the lingering temp files are removed and for any further issues.
